platform/ios/src/MGLMapView.mm

```diff
@@ -4150,6 +4173,111 @@ - (void)applyIconIdentifier:(NSString *)iconIdentifier toAnnotationsWithImageReu
     }
 }
 
+#pragma mark - Shape Annotation
+
+- (void)selectShapeAnnotation:(MGLShape *)shapeAnnotation
+{
+    if (!shapeAnnotation) return;
+    
+    if (shapeAnnotation == _selectedShapeAnnotation) return;
+    
+    [self deselectShapeAnnotation:_selectedShapeAnnotation];
+    
+    _selectedShapeAnnotation = shapeAnnotation;
+    
```

Comment:
The intent of #2082 was that we’d display a callout over the selected shape. Per https://github.com/mapbox/mapbox-gl-native/issues/2082#issuecomment-229487695, we’d need to constrain the coordinate to the current viewport before calculating the centroid. So `coordinate` wouldn’t give us quite the right behavior, but we could fashion a method that takes an `MGLCoordinateBounds` and clips the polygon to polyline to that bounds before passing the geometry off to Polylabel.

Reply:
> The intent of #2082 was that we’d display a callout over the selected shape.

Our [Android SDK gained essentially the exact same functionality](https://github.com/mapbox/mapbox-gl-native/issues/2082#issuecomment-313594729) that is proposed here in  https://github.com/mapbox/mapbox-gl-native/pull/9443. I don't think there is any mechanism in that SDK to assist with callout placement, yet. I think it would be ok to use the same approach and keep #2082 open and handle callouts separately.

Reply:
I guess that would be an argument for adding a separate set of delegate methods, because it would be weird for `-mapView:didSelectAnnotation:` to work with `-mapView:annotationCanShowCallout:` in some cases but not others.

It doesn’t feel great to potentially have a point annotation and shape annotation both be selected at the same time. It’ll lock us into a design that’s reminiscent of MapKit but quite different, even after we implement the callout views.

Reply:
My principal deterrent to avoid using `[MGLMapView annotationTagAtPoint:persistingResults:]` was the fact that *`MGLShape` "annotations" does not have a callout view. So selecting an annotation that handles a callout only in specific cases would be confusing for developers and interpreted as a bug.

But selecting a `MGLShape` annotation deselects `MGLPointAnnotation` and viceversa.

Reply:
I think the broader question is whether shapes should have callouts. The original request in #2082 and much of the subsequent comments were about callouts, so I guess I’d consider that PR unfinished without them. If we land this feature without callouts and with an API that’s separate from normal annotation selection, does that make it more difficult for us to implement shape callouts later? Or are we counting on being able to clean up this API in v4.0.0?

Reply:
Right, the original request in #2082 was callouts. I should have referred to #3720 which is what the Android implementation https://github.com/mapbox/mapbox-gl-native/pull/9443 fixes. 

Maybe we could wait and refactor our Annotation API, then support callouts for shape annotations, this is how it looks right know so far.
<img width="1019" alt="annotation_api" src="https://user-images.githubusercontent.com/2745166/30165014-7af72792-93ac-11e7-8c3b-8ce798006f9a.png">

Reply:
How much of a lift would it be to support callouts for shape annotations? At a glance, it looks like most of the pieces are already in place (especially polylabel).